frame/bounties/src/lib.rs

```diff
@@ -722,6 +1521,114 @@ impl<T: Config> Module<T> {
+
+		let bounty = Self::bounties(&bounty_id).ok_or(Error::<T>::InvalidIndex)?;
+
+		if let BountyStatus::Active { curator, update_due } = bounty.status {
+			Ok(Some((curator, update_due)))
+		} else {
+			if is_reject_origin {
+				// If call originates from T::RejectOrigin
+				// ignore state check on parent bounty,
+				// and force execution.
+				Ok(None)
+			} else {
+				Err(Error::<T>::UnexpectedStatus.into())
+			}
+		}
+	}
+
+	fn create_subbounty(
+		bounty_id: BountyIndex,
+		subbounty_id: BountyIndex,
+		description: Vec<u8>,
+	) {
+
+		let subbounty = SubBounty {
+			fee: 0u32.into(),
+			curator_deposit: 0u32.into(),
+			status: SubBountyStatus::Added,
+		};
+		SubBounties::<T>::insert(bounty_id, subbounty_id, &subbounty);
+		BountyDescriptions::insert(subbounty_id, description);
+		Self::deposit_event(RawEvent::SubBountyAdded(bounty_id, subbounty_id));
+	}
+
+	fn impl_close_subbounty(
+		bounty_id: BountyIndex,
+		subbounty_id: BountyIndex,
+	) -> DispatchResult {
+		SubBounties::<T>::try_mutate_exists(
+			bounty_id,
+			subbounty_id,
+			|maybe_subbounty| -> DispatchResult {
+
+				let subbounty = maybe_subbounty
+					.as_mut()
+					.ok_or(Error::<T>::InvalidIndex)?;
+
+				match &subbounty.status {
+					SubBountyStatus::Added |
+					SubBountyStatus::SubCuratorProposed { .. } => {
+						// Nothing extra to do besides the removal of the subbounty below.
+					},
+					SubBountyStatus::Active { subcurator } => {
+						// Cancelled by admin(master curator or Root origin),
+						// refund deposit of the working subcurator.
+						let _ = T::Currency::unreserve(subcurator, subbounty.curator_deposit);
+						// Then execute removal of the subbounty below.
+					},
+					SubBountyStatus::PendingPayout { .. } => {
+						// Subbounty is already in pending payout. If admin(master curator or
+						// Root origin) wants to cancel this subbounty,
+						// it should mean the subcurator was acting maliciously.
+						// So admin should first unassign the subcurator,
+						// slashing their deposit.
+						return Err(Error::<T>::PendingPayout.into())
+					},
+				}
+
+				// revert the subcurator fee back to master curator &
+				// Reduce the active subbounty count.
+				Bounties::<T>::mutate_exists(
+					bounty_id,
+					|maybe_bounty| {
+						if let Some(bounty) = maybe_bounty.as_mut() {
+							bounty.fee = bounty
+								.fee
+								.saturating_add(subbounty.fee);
+							bounty.active_subbounty_count = bounty.active_subbounty_count
+								.saturating_sub(1);
+						}
+					}
+				);
+
+				// Transfer fund from subbounty to parent bounty.
+				let bounty_account = Self::bounty_account_id(bounty_id);
+				let subbounty_account = Self::bounty_account_id(subbounty_id);
+				let balance = T::Currency::free_balance(&subbounty_account);
+				let _ = T::Currency::transfer(
+					&subbounty_account,
+					&bounty_account,
+					balance,
+					AllowDeath
+				); // should not fail
```

Comment:
likewise, this process also seem to be duplicated in a few places, can optionally be made into a helper.

Reply:
Good Suggestion, Sure I agree, there are few duplicates common between Bounties and Subbounties. I am not supposed to change Bounties code as part of the PR, Wish to take the proposed code optimization as part of separate PR.